### ISSUE TITLE
Add a way to download the original source code

### ISSUE THREAD

**Issue by snake-py** [snake-py/eject#4]
Right now the cli will simply copy the files from `node_modules`, the next step is that we are able to download the files from the repository. Then the user can make changes or directly import them.

**Comment by snake-py** [comment]
I have two main ideas on how to do it, depending on the later development flow. But some things to consider:

1. The import paths might change due to aliasing in the original code and different package.json structure
2. The library which gets ejected might have a complex bundler config. 

These two points bring me to the conclusion that we need to prebundle the library with the original bundler configs. So that devs still import from bundled files. The downside of this approach is obviously longer wait times and it feels not very smoothly. 

I guess the smoothes way would be that devs can import files directly from the source code, however how can we then map imports? I think this is where we need to have library maintainers define a source code mapping to their package.json exports. Or we offer lib maintainers to add first class support by letting them add a custom eject script.

**Comment by brillout** [comment]
Your analysis makes perfect sense.

I think, eventually, it would be nice to support both levels.
- Level 2: The library's `build` script is also ejected and used by the user.
  - Eject would basically modify the user's `package.json` like this:
    ```diff
      {
        "scripts": {
    -     "build": "echo 'the original build script'"
    +     "build": "npm run build:eject && npm run build:app",
    +     "build:app": "echo 'the original build script'"
    +     "build:ejected": "(cd ../ejected/some-library/ && npm run build) && (cd ../ejected/some-other-library/ && npm run build)"
        }
      }
    ```
- Level 3: Build-less eject: no `package.json` nor `tsconfig.json` is ejected. In other words, no `file:` linking and the user's app directly uses ejected files instead.
  - For Vike extensions, the import paths of user code don't need to be modified, because the ejected source are Vike entries (i.e. `+` files).
  - I think you're right, the library needs to specify which files and where these files should be ejected.
  - For a library to support level 3, its `tsconfig.json` needs to be strict (`"strict": true` and `"noUncheckedIndexedAccess": true`). So that the library's typescript is compatible with the user's `tsconfig.json`.

Maybe skipping level 2 and only supporting level 3 would be easier for now :thinking:

**Comment by brillout** [comment]
How about we skip level-2 and directly try level-3 instead? So far I think level-2 isn't worth it.

How about something like [this](https://github.com/brillout/playground_eject-vike-react-chakra) for level-3?

The only question left is how can we remove `vike-react-chakra` from `extends`? Is there a way to do it that doesn't require AST parsing? Maybe we can do something hackish for now.

**Comment by brillout** [comment]
> How about something like [this](https://github.com/brillout/playground_eject-vike-react-chakra) for level-3?

I went ahead and added [`eject.config.js`](https://github.com/vikejs/vike-react/blob/main/packages/vike-react-chakra/eject.config.js), in case you want to play with it.

Note that it isn't included in the `vike-react-chakra` npm package because it doesn't need to: the source code will be fetched and will include `eject.config.js`.

**Comment by snake-py** [comment]
So I think there is maybe a difference in my head let's clarify this first:

eject => ejecting a dependency as is (either sourced or bundled), if ejecting source code imports should be fixed. 
scaffold => placing files directly into the project source code

I think we should not mix these up. To me this is a two part process.

I would first now work on downloading the source files. Next something like you are proposing [here](https://github.com/vikejs/vike-react/blob/main/packages/vike-react-chakra/eject.config.js) makes sense. 


> The only question left is how can we remove vike-react-chakra from extends? Is there a way to do it that doesn't require AST parsing? Maybe we can do something hackish for now.

I think this is basically a question for later, no? I think you are looking at this from the external lib maintainer perspective, which makes sense but maybe at a later stage? I guess here we now want first to build a way that a dependency can add custom scaffold actions to perform. 

Possible actions a lib want to execute for the scaffolding:

- move/copy files
- updated config files 
- run some custom script
- add some dependencies


## Proposal

My proposal is basically the following:

- add a flag which decides whether to eject using source or bundle (default will be source?)
- detect if the ejected source code holds a eject.configs.js(on)
- add a flag to directly scaffold
- add a new command to scaffold already ejected projects 

The config:

```js
// node_modules/vike-react-chakra/eject.config.js

export const config = {
  eject: {
  // maybe configs which are required for fixing aliases and imports
  },
  scaffold: {
  operations: ['mv Wrapper.tsx pages/+Wrapper.tsx'], // I think it might be better so simply have them define a script for now
  // script: // path to node script ?
 }
}
```

### New look


```
npx eject <dependency> // downloads source code and fixes imports
npx eject <dependency> --scaffold // downloads and scaffold dependency if already downloaded only scaffold
npx eject <dependency> --bundle // how it currently works
```

**Comment by snake-py** [comment]
> The vast majority of libraries don't use path aliases anyways.

I am tbh, not so sure about that e.g. https://github.com/mui/material-ui/blob/master/packages/mui-base/src/Menu/Menu.test.tsx

**Comment by snake-py** [comment]
Also, something to think about if we are ejecting source code. Should they be a mechanic to pull, then it would make sense to create it as git-sub-module, no?

**Comment by brillout** [comment]
> > The vast majority of libraries don't use path aliases anyways.
> 
> I am tbh, not so sure about that e.g. https://github.com/mui/material-ui/blob/master/packages/mui-base/src/Menu/Menu.test.tsx

I don't think it's using path aliases; it seems to be a monorepo. See https://www.npmjs.com/package/@mui/internal-test-utils and https://www.npmjs.com/package/@mui/base. I ain't aware of any library using path aliases. If a library where to use path aliases then it would need to resolve them all at build time anyways: a library enforcing a path alias to its users is a no-go. I'd suggest we worry about path aliases only after we encounter a problem (although I'm fairly confident we won't :wink:).

**Comment by brillout** [comment]
> Also, something to think about if we are ejecting source code. Should they be a mechanic to pull, then it would make sense to create it as git-sub-module, no?

Yea, I thought about this. It could be neat to have "live" ejections. I'm inclined to think we should work on this later because managing Git repositories within Git repositories is complex (maybe the simpler solution will be to eject the library outside of the user's Git repository). That said, it would be really nice also because it will nudge users to contribute. But quite complex so let's postpone this, I think.

**Comment by brillout** [comment]
> My proposal

Makes sense!

Also one thing that will be important (even more so than live ejections I think) is "partial eject" (or partial scaffold to re-use your term) where only some functionality of the library is ejected. Note that the example I gave is ejecting the entire functionality of `vike-react-chakra` (despite ejecting only one file).

I'll wrote more about your proposal later today.

**Comment by snake-py** [comment]
> Yea, I thought about this. It could be neat to have "live" ejections. I'm inclined to think we should work on this later because managing Git repositories within Git repositories is complex (maybe the simpler solution will be to eject the library outside of the user's Git repository). That said, it would be really nice also because it will nudge users to contribute. But quite complex so let's postpone this, I think.

I think it is not too complex. Basically, it is just setting up a [submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules). I have done this quite often now. 

In addition, since git is a requirement, I guess the simplest way to download the source code will even be a [shallow clone](https://stackoverflow.com/questions/11497457/git-clone-without-git-directory).

**Comment by brillout** [comment]
:+1: Ok sounds good. (Personally I've always regretted and stopped using submodules, but maybe it's just me.)

**Comment by brillout** [comment]
> > I would first now work on downloading the source files. Next something like you are proposing [here](https://github.com/vikejs/vike-react/blob/main/packages/vike-react-chakra/eject.config.js) makes sense.

:+1: We can also work on level-2 before starting to work on level-3.

> add a flag

I agree we will need a flag or a CLI argument for partial eject.

But, for full ejection AFAICT we can have a sensible default instead of a flag. In other words, how about we try the highest level first:
- If the library defines an `eject.config.js` => level-3, that is TypeScript ejection without build script => eject picks this level if it's available.
- If the library supports level-2 (TypeScript ejection with build script) => eject picks this level if it's available.
- Fallback to level-1 (JavaScript `node_modules/` ejection) if level-2 and level-3 aren't available.

AFACIT level-3 is better than than level-2 (and level-2 is better than level-1) so I don't think we need a flag. Or maybe just a flag if there is use case for overriding eject's default behavior.

> I think it might be better so simply have them define a script for now

:+1: Good idea. We could then/eventually provide some commonly needed utilities.

**Comment by brillout** [comment]
> add a new command to scaffold already ejected projects

Interesting idea. I wonder how often users would need that. Do you have a concrete use case in mind?

**Comment by snake-py** [comment]
> Interesting idea. I wonder how often users would need that. Do you have a concrete use case in mind?

In case you eject a library and then later on you notice in the docs that you can also scaffold. Basically, it is like a user will do:

1. "Ah I can eject the thing" => runs eject
2. "Oh the lib supports scaffolding" => wants to scaffold already ejected lib

Alternative

1. User has ejected lib (with git connection)
2. Maintainer adds scaffolding
3. User wants to scaffold newly added stuff

**Comment by snake-py** [comment]
> I agree we will need a flag or a CLI argument for partial eject.

You mean partial scaffold? 

I think we could have libs define multiple scaffold commands themselfs:



```js
export const config = {
  eject: {
  // maybe configs which are required for fixing aliases and imports
  },
  scaffold: {
  default: 'defaultScript.js',
  wrapper: 'scaffoldWrapper.js'
 }
}
````


````
npx eject <dependency> --scaffold // would run default
npx eject <dependency> --scaffold --wrapper // would run wrapper
npx eject <dependency> --scaffold --list (would list the available options requirement source code is ejected)
````


Btw as an alternative. The scaffolding scripts could even just be npm scripts in the package and a user could run these themselfs

**Comment by brillout** [comment]
> scaffold already ejected lib

Interesting idea. I like it and it makes sense: if the ejection is live then the user is essentially using the full-blown thing, it's just that it's linked locally and potentially forked in some areas.

> You mean partial scaffold?

I mean ejecting/scaffolding only *some* functionalities of the package. Imagine a component library and you eject/scaffold only one component e.g. the date picker (instead of ejecting all components of the library).

How about this:

```
npx eject <dependency> // would run default
npx eject <dependency> wrapper // would run wrapper
npx eject <dependency> --list (would list the available options requirement source code is ejected)
```

I'd argue it's a little simpler. As for the using the "scaffold" term I think it would add more confusion than add benefits: it will be difficult to teach users the distinction between "eject" and "scaffold". Whereas "partial eject" is self explanatory, I think.

**Comment by snake-py** [comment]
There is one final thing which I would like to discuss before I implement the changes. 

Basically, your end goal is for `vike` and any `vike-packages`, to be able to partially eject code. Here is the thing, have you also thought of it from a code generation perspective like many other frameworks do it?

For instance, in Laravel you can do:

````bash
php artisan make:controller MyController // will create a controller
php artisan make:controller MyController --ressources // will create a controller with some best practices
````

Wouldn't it make more sense to have a tool like this for vike?

This would mean any vike package can add commands to the `make` directive. 

```
npm i vike-react // installs vike-react and inside the package there is a command part which registers new commands to the vike cli

vike make:{file/component} 
vike make:wrapper Wrapper --react 
````

This would basically solve the issue from another angel.

**Comment by brillout** [comment]
Correct me if I'm wrong, but the scaffolding feature of other frameworks are only about generating boilerplate code. An example would be `$ vike add page` that would prompt a couple of questions and then generate `pages/a-new-route/*` files. While it's a nice-to-have I don't think it adds *that much* value. It's just a nice shortcut, nothing more.

Ejecting a Vike extension is a whole different beast. The goal here is to enable users to gain control over tool integration (which is key aspect of architectural control).

For example authentication. A generic authentication implementation can work for many users. But many other users, which have more or less unique authentication requirements, need extra control. So the idea here is to have a Vike extension `vike-auth` that works for most use cases, while enabling users to gain control by ejecting. Instead of ejecting the whole extension at once, the user can cherry-pick the parts of the authentication implementation he wants to customize. For example, if he wants to customize the authentication pages (`/login`, `/reset`, ...) then he can `$ eject vike-auth user-interface`: all the UI components are then moved to his Git repository &mdash; while still using the authentication business logic defined by `vike-auth` (which isn't ejected). That's one example with what I mean with "partial eject".

The long term vision is to have all glue code be initially hidden to users, while the user can eject the generic glue code doesn't work for him.

We've started going in that direction: see for example `For full customization consider [ejecting](https://vike.dev/eject)` at https://github.com/vikejs/vike-react/tree/main/packages/vike-react-antd#:~:text=For%20full%20customization%20consider%20ejecting

Another way to put it:
 - A given scaffold command (e.g. `$ vike add page`) is usually called several times.
 - A given eject command (e.g. `$ eject vike-auth user-interface`) is usually called only once. (It doesn't really make sense to call it again since the code is already ejected.)

While scaffolding is nice, I think eject is a feature that is a lot more powerful. Thus I'm inclined to prioritize eject over scaffolding.

**Comment by snake-py** [comment]
Okay, I see the reasoning, but then the question is why should eject place source files outside of the eject folder? The user can just go into the ejected source code make the required changes there?

Why deal with something like this:
> operations: ['mv Wrapper.tsx pages/+Wrapper.tsx'],

The user can just go and modify `ejected/{package}/pages/Wrapper.tsx`, no?

**Comment by snake-py** [comment]
> The only question left is how can we remove vike-react-chakra from extends? Is there a way to do it that doesn't require AST parsing? Maybe we can do something hackish for now.

Also I have been thinking about this. I guess I need some more context, since I have not used alot of vike extensions. How is extend working internally? Is it pushing simply all components onto each other or is it somehow merging and overriding?

E.g. if I extend both of these: [vike-react](https://vike.dev/vike-react), [vike-react-query](https://github.com/vikejs/vike-react/tree/main/packages/vike-react-query#readme) and both define a Wrapper, what Wrapper would I use? What happens if I in addition define my own Wrapper?


Edit:

Maybe vike-react and vike-react-query are bad examples. Basically I mean two extensions defining same components

**Comment by brillout** [comment]
Yes, from a functional perspective, you're right it's the same.

It's just about removing unnecessary boilerplate.

Take for example [`vike-react-chakra`](https://github.com/vikejs/vike-react/tree/main/packages/vike-react-chakra): the integration is [just a `+Wrapper`](https://github.com/vikejs/vike-react/blob/main/packages/vike-react-chakra/Wrapper.tsx). So instead of having `ejected/vike-react-chakra/{pacakge.json,config.ts,Wrapper.tsx,CHANGELOG.md,README.md,...}` it could simply eject the wrapper at `pages/+Wrapper.tsx`.

**Comment by snake-py** [comment]
> Yes, from a functional perspective, you're right it's the same.

I think then we want for now just this no? I think you called it `level 2` support?

**Comment by brillout** [comment]
> > The only question left is how can we remove vike-react-chakra from extends? Is there a way to do it that doesn't require AST parsing? Maybe we can do something hackish for now.
> 
> Also I have been thinking about this. I guess I need some more context, since I have not used alot of vike extensions. How is extend working internally? Is it pushing simply all components onto each other or is it somehow merging and overriding?

It just merges both configs. (More precisely speaking [it extends configs](https://vike.dev/extends)).

Btw. I talked to Joël (the author of Bati; he has a lot of experience with these kind of things) and, apparently, it's easy to do with https://github.com/benjamn/recast.

> E.g. if I extend both of these: [vike-react](https://vike.dev/vike-react), [vike-react-query](https://github.com/vikejs/vike-react/tree/main/packages/vike-react-query#readme) and both define a Wrapper, what Wrapper would I use? What happens if I in addition define my own Wrapper?

I think we can do it with recast: we rename the file to `pages/WrapperChakra.ts` (to avoid overriding any existing file) and then import it in `pages/+config.js` (instead of making it a `+` file, see https://vike.dev/config).

> for now just this

Yea, we can also only work on the git cloning approach for now.

**Comment by snake-py** [comment]
Maybe interesting to note. I am maintaining a component library with path aliases (on my job) and I am planing to make the lib ejectable using a git submodule approach. I think I found a solid way to do it. But it brought me to the question, why having this cli benefits, if any user could just add the lib they need as sub module.

**Comment by brillout** [comment]
It's about making it easier and incentivizing users to do it.

Also partial eject will play a central role in Vike's future. Doing partial eject manually is cumbersome.

**Comment by snake-py** [comment]
SO I thought about how I will go about implementing this. One thing I am unsure about is if the package is withing a monorepo.

Basically I thought, that we want to use git based on a link in package.json. But the more I think about it, I guess we want to download a zip based on a download path? However, this will make "life ejections" harder 🤔 ? 

In anyway I have time this week and will get started on this.

**Comment by snake-py** [comment]
To summarize the whole discussion:

We need a tool which can do the following:

- download the whole source code and place it into ejected (full ejection)
  - follow up on this it might connect this to a git repository (life ejections)
- download only part of the source code and place it directly into the app and leave the dependency tree untouched (probably better to define this through a config?)

**Comment by brillout** [comment]
Exactly. Looking forward to it!

I wonder what the path of least resistance will be. So far it seems that full-ejection (without live connection) is the easiest :eyes: let's see.

**Comment by snake-py** [comment]
So I started and I have the first question. What do you think makes more sense for downloading the source files:

- use git to clone the repository 
  - in case for monorepos I guess some special logic is required (like deleting everything except the one package, I guess this could be done by the script which is placed within the source code of the package?)
- use a download function to download a zip/file directly (I think this is more flexible, but also more challenging or rather has more prerequisites) 

In case for downloading the zip:

Here again monorepos are a little more challenging. In case it is not a monorepo we can simply just use a source eg. `https://github.com/vikejs/vike-react/archive/refs/heads/main.zip`. In case for monorepos we could then copy only a specific path from the zip. 

Any idea for an approach here?

**Comment by brillout** [comment]
How about `$ git clone` then we remove `.git` and all other irrelevant files?

For example, remove all files that aren't part of this directory: https://github.com/vikejs/vike-react/blob/66d083838624dc9908e7b75c1eef8bbc0b6da3d7/packages/vike-react-chakra/package.json#L48.

The nice thing about using Git is that more than 99% of users already have Git installed. So no annoying extra system dependency is required.

**Comment by brillout** [comment]
> I guess this could be done by the script which is placed within the source code of the package?

So far, I don't think we need this.

**Comment by brillout** [comment]
I think https://github.com/vikejs/vike-react/blob/66d083838624dc9908e7b75c1eef8bbc0b6da3d7/packages/vike-react-chakra/package.json#L48 is all we need to make it work.
